cpp/include/cudf/groupby.hpp

```diff
@@ -0,0 +1,219 @@
+   *
+   * If `false`, rows in the `keys` table with NULL values will be treated as
+   * any other row. Furthermore, a NULL value will be considered equal to
+   * another NULL value. For example, two rows `{1, 2, 3, NULL}` and `{1, 2, 3,
+   * NULL}` will be considered equal, and their associated rows in the `values`
+   * table will be aggregated.
+   *
+   * @note The behavior for a Pandas groupby operation is `ignore_null_keys ==
+   *true`.
+   * @note The behavior for a SQL groupby operation is `ignore_null_keys ==
+   *false`.
+   *
+   *---------------------------------------------------------------------------**/
+  bool const ignore_null_keys{true};
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief Supported aggregation operations
+ *
+ *---------------------------------------------------------------------------**/
+enum operators { SUM, MIN, MAX, COUNT, MEAN, MEDIAN, QUANTILE};
+
+namespace hash {
+
+/**---------------------------------------------------------------------------*
+ * @brief  Options unique to the hash-based groupby
+ *---------------------------------------------------------------------------**/
+struct Options : groupby::Options {
+  Options(bool _ignore_null_keys = true)
+      : groupby::Options(_ignore_null_keys) {}
+}; 
+
+/**---------------------------------------------------------------------------*
+ * @brief Performs groupby operation(s) via a hash-based implementation
+ *
+ * Given a table of keys and corresponding table of values, equivalent keys will
+ * be grouped together and a reduction operation performed across the associated
+ * values (i.e., reduce by key). The reduction operation to be performed is
+ * specified by a list of operator enums of equal length to the number of value
+ * columns.
+ *
+ * The output of the operation is the table of key columns that hold all the
+ * unique keys from the input key columns and a table of aggregation columns
+ * that hold the specified reduction across associated values among all
+ * identical keys.
+ *
+ * @param keys The table of keys
+ * @param values The table of aggregation values
+ * @param ops The list of aggregation operations
+ * @return A tuple whose first member contains the table of output keys, and
+ * second member contains the table of reduced output values
+ *---------------------------------------------------------------------------**/
+std::pair<table, table> groupby(table_view const& keys,
+                                table_view const& values,
+                                std::vector<operators> const& ops,
+                                Options options = Options{});
+
+}  // namespace hash
+
+namespace sort {
+
+/**---------------------------------------------------------------------------*
+ * @brief  Arguments to be passed to an operator
+ *---------------------------------------------------------------------------**/
+struct operation_args {};
```

Comment:
@jrhemstad - opinion on this?

Reply:
The problem is that the `hash` groupby does not and can not support things like `quantile_args` because a hash groupby cannot be used to compute a quantile.

We've got some plans for overhauling the structure of groupby, so I'd suggest just keeping the idiosyncratic APIs for now.